### PR TITLE
[GEOT-5170] A visitor to optimize filters for in memory evaluation

### PR DESCRIPTION
**Pull request by aaime**
See: https://osgeo-org.atlassian.net/browse/GEOT-5170

### CODE REVIEW

**Comment by jodygarnett** [comment]
I will review and merge, need to add a note to the docs.

**Comment by bencaradocdavies** [comment]
Hmm, I cherry-picked this onto master and see failures in gt-jdbc-postgis:

Failed tests: 
  PostgisTemporalFilterOnlineTest>OnlineTestCase.run:123->JDBCTemporalFilterOnlineTest.testTContains:178->JDBCTemporalFilterOnlineTest.assertDatesMatch:70->JDBCTemporalFilterOnlineTest.assertDatesMatch:76 expected:<2> but was:<0>
  PostgisTemporalFilterOnlineTest>OnlineTestCase.run:123->JDBCTemporalFilterOnlineTest.testTContains:178->JDBCTemporalFilterOnlineTest.assertDatesMatch:70->JDBCTemporalFilterOnlineTest.assertDatesMatch:76 expected:<2> but was:<0>

**Comment by bencaradocdavies** [comment]
I can confirm that the gt-jdbc-postgis test fails on master if and only if this change is applied.

I have not yet tested gs-app-schema-test (offline or online).

**Comment by bencaradocdavies** [comment]
@aaime, @jodygarnett, the good news is that, with this change on master, a full GeoServer build passes (-Prelease), and GeoServer app-schema online tests pass against postgis. (We have previously had some complex feature fun and games after filter optimisation (SimplifyingFilterVisitor MatchAction), so I thought I should check.)

The only failure I could find is the gt-jdbc-postgis online test failure described above.

**Comment by jodygarnett** [comment]
Given that @aaime is doing way too much this week are you in position to look at the failure reported above @bencaradocdavies ?

I am still trying to sort out where this goes in our docs, looking at (https://osgeo-org.atlassian.net/browse/GEOT-5170) and the code it looks like this class is intended to be extended as an optimization.  Sound like it should be covered in the section on the different visitor implementations, possibly with a code example (probably for Date which always seems to be the problem when sorting).

**Comment by bencaradocdavies** [comment]
@jodygarnett I had a look but could not see the problem. Some type conversion problem that occurs in jdbc or postgis, perhaps? I have not yet stepped through it in the debugger and may not have time today.

@ianturton if you have a moment could you please advise? You may have seen similar jdbc or postgis date handling problems in your recent fixes.

File: modules/library/main/src/main/java/org/geotools/filter/visitor/BindingFilterVisitor.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        Expression expr2 = optimizeTime(overlappedBy.getExpression2(), extraData);
+
+        return getFactory(extraData).overlappedBy(expr1, expr2, overlappedBy.getMatchAction());
+    }
+
+    public Object visit(TContains contains, Object extraData) {
+        Expression expr1 = optimizeTime(contains.getExpression1(), extraData);
+        Expression expr2 = optimizeTime(contains.getExpression2(), extraData);
+
+        return getFactory(extraData).contains(expr1, expr2, contains.getMatchAction());
```
Comment (bencaradocdavies) [inline modules/library/main/src/main/java/org/geotools/filter/visitor/BindingFilterVisitor.java]: Aha! Found it. This statement should use the method "tcontains" not "contains". Changing this method to "tcontains" fixes the build failure in gt-jdbc-postgis online tests.

**Comment by bencaradocdavies** [comment]
@aaime @jodygarnett please see #935 for a new pull request that should replace this one. #935 is Andrea's commit (cherry-picked) and my tcontains fix.

**Pull request closed without merging**